File: backend/src/infrastructure/mangadex/mangadex.client.ts
import { env } from "../../shared/configs/app.config.js";
import { HttpError } from "../../shared/errors/http-error.js";
import type { ChapterSummary, MangaDexListResponse, MangaDexSingleResponse, MangaSummary, ReaderPayload } from "./mangadex.types.js";

const USER_AGENT = "mangadex-reader/0.1 (+https://github.com/local/mangadex-reader)";
const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

export type MangaSearchSort = "relevance" | "latest" | "followed" | "title" | "created" | "updated";

export type MangaSearchInput = {
  q?: string;
  limit: number;
  offset: number;
  languages: string[];
  tags: string[];
  includedTags?: string[];
  excludedTags?: string[];
  contentRating?: string[];
  status?: string[];
  year?: number;
  demographic?: string[];
  sort?: MangaSearchSort;
};

export function getMangaDexUploadsBaseUrl() {
  if (env.MANGADEX_UPLOADS_BASE_URL) return env.MANGADEX_UPLOADS_BASE_URL;
  return env.MANGADEX_BASE_URL.includes(".dev") ? "https://uploads.mangadex.dev" : "https://uploads.mangadex.org";
}

function firstLocalized(value: Record<string, string> | undefined, preferred = ["vi", "en", "ja-ro", "ja"]) {
  if (!value) return "";
  for (const lang of preferred) {
    if (value[lang]) return value[lang];
  }
  return Object.values(value)[0] ?? "";
}

function appendArrayParams(params: URLSearchParams, key: string, values: string[]) {
  for (const value of values) {
    params.append(`${key}[]`, value);
  }
}

async function requestMangaDex<T>(path: string, params?: URLSearchParams): Promise<T> {
  const url = new URL(path, env.MANGADEX_BASE_URL);
  if (params) {
    url.search = params.toString();
  }

  for (let attempt = 0; attempt < 2; attempt += 1) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10_000);
    try {
      const response = await fetch(url, {
        headers: {
          Accept: "application/json",
          "User-Agent": USER_AGENT
        },
        signal: controller.signal
      });

      if (response.status === 429 && attempt === 0) {
        await new Promise((resolve) => setTimeout(resolve, 750));
        continue;
      }

      if (!response.ok) {
        throw new HttpError(response.status, `MangaDex request failed: ${response.statusText}`, "MANGADEX_ERROR");
      }

      return (await response.json()) as T;
    } catch (error) {
      if (error instanceof HttpError) {
        throw error;
      }
      if (attempt === 0 && error instanceof Error && error.name === "AbortError") {
        continue;
      }
      if (error instanceof Error) {
        throw new HttpError(502, `Unable to reach MangaDex: ${error.message}`, "MANGADEX_UNREACHABLE");
      }
      throw error;
    } finally {
      clearTimeout(timeout);
    }
  }

  throw new HttpError(504, "MangaDex request timed out", "MANGADEX_TIMEOUT");
}

function normalizeManga(entity: MangaDexSingleResponse["data"]): MangaSummary {
  const attrs = entity.attributes as Record<string, unknown>;
  const cover = entity.relationships?.find((relationship) => relationship.type === "cover_art");
  const fileName = cover?.attributes?.fileName;
  const tags = ((attrs.tags as Array<{ attributes?: { name?: Record<string, string> } }>) ?? [])
    .map((tag) => firstLocalized(tag.attributes?.name, ["en", "vi"]))
    .filter(Boolean);
  const altTitles = ((attrs.altTitles as Array<Record<string, string>>) ?? [])
    .map((title) => firstLocalized(title))
    .filter(Boolean)
    .slice(0, 8);

  return {
    id: entity.id,
    title: firstLocalized(attrs.title as Record<string, string> | undefined),
    altTitles,
    description: firstLocalized(attrs.description as Record<string, string> | undefined),
    status: attrs.status as string | undefined,
    year: attrs.year as number | undefined,
    contentRating: attrs.contentRating as string | undefined,
    tags,
    coverUrl: typeof fileName === "string" ? `${getMangaDexUploadsBaseUrl()}/covers/${entity.id}/${fileName}.512.jpg` : undefined
  };
}

function normalizeChapter(entity: MangaDexListResponse["data"][number]): ChapterSummary {
  const attrs = entity.attributes as Record<string, unknown>;
  const group = entity.relationships?.find((relationship) => relationship.type === "scanlation_group");
  return {
    id: entity.id,
    title: (attrs.title as string | null) ?? "",
    chapter: (attrs.chapter as string | null) ?? null,
    volume: (attrs.volume as string | null) ?? null,
    translatedLanguage: (attrs.translatedLanguage as string | undefined) ?? "en",
    publishAt: (attrs.publishAt as string | undefined) ?? "",
    pages: (attrs.pages as number | undefined) ?? 0,
    scanlationGroup: group?.attributes?.name as string | undefined
  };
}

export async function searchManga(input: {
  q?: string;
  limit: number;
  offset: number;
  languages: string[];
  tags: string[];
  includedTags?: string[];
  excludedTags?: string[];
  contentRating?: string[];
  status?: string[];
  year?: number;
  demographic?: string[];
  sort?: MangaSearchSort;
}) {
  const includedTags = [...new Set([...(input.tags ?? []), ...(input.includedTags ?? [])])].filter((tag) => UUID_PATTERN.test(tag));
  const excludedTags = (input.excludedTags ?? []).filter((tag) => UUID_PATTERN.test(tag));
  const params = new URLSearchParams({
    limit: String(input.limit),
    offset: String(input.offset)
  });

  applyMangaSort(params, input.sort ?? (input.q ? "relevance" : "followed"));
  if (input.q) params.set("title", input.q);
  if (input.year) params.set("year", String(input.year));
  appendArrayParams(params, "availableTranslatedLanguage", input.languages);
  appendArrayParams(params, "includes", ["cover_art"]);
  appendArrayParams(params, "contentRating", input.contentRating?.length ? input.contentRating : ["safe", "suggestive"]);
  appendArrayParams(params, "status", input.status ?? []);
  appendArrayParams(params, "publicationDemographic", input.demographic ?? []);
  appendArrayParams(params, "includedTags", includedTags);
  appendArrayParams(params, "excludedTags", excludedTags);

  const result = await requestMangaDex<MangaDexListResponse>("/manga", params);
  return {
    limit: result.limit,
    offset: result.offset,
    total: result.total,
    data: result.data.map(normalizeManga)
  };
}

export function applyMangaSort(params: URLSearchParams, sort: MangaSearchSort) {
  if (sort === "latest") {
    params.set("order[latestUploadedChapter]", "desc");
    return;
  }
  if (sort === "title") {
    params.set("order[title]", "asc");
    return;
  }
  if (sort === "created") {
    params.set("order[createdAt]", "desc");
    return;
  }
  if (sort === "updated") {
    params.set("order[updatedAt]", "desc");
    return;
  }
  if (sort === "relevance") {
    params.set("order[relevance]", "desc");
  }
  params.set("order[followedCount]", "desc");
}

export async function getManga(id: string) {
  const params = new URLSearchParams();
  appendArrayParams(params, "includes", ["cover_art"]);
  const result = await requestMangaDex<MangaDexSingleResponse>(`/manga/${id}`, params);
  return normalizeManga(result.data);
}

export async function getChapters(input: { mangaId: string; limit: number; offset: number; translatedLanguage: string[] }) {
  const params = new URLSearchParams({
    limit: String(input.limit),
    offset: String(input.offset),
    "order[volume]": "asc",
    "order[chapter]": "asc"
  });
  appendArrayParams(params, "translatedLanguage", input.translatedLanguage);
  appendArrayParams(params, "includes", ["scanlation_group"]);

  const result = await requestMangaDex<MangaDexListResponse>(`/manga/${input.mangaId}/feed`, params);
  return {
    limit: result.limit,
    offset: result.offset,
    total: result.total,
    data: result.data.map(normalizeChapter)
  };
}

export async function getReader(chapterId: string): Promise<ReaderPayload> {
  const result = await requestMangaDex<{
    baseUrl: string;
    chapter: { hash: string; data: string[]; dataSaver: string[] };
  }>(`/at-home/server/${chapterId}`);

  const { baseUrl, chapter } = result;
  return {
    baseUrl,
    hash: chapter.hash,
    pages: chapter.data,
    dataSaverPages: chapter.dataSaver,
    pageUrls: chapter.data.map((page) => `/api/pages/${chapterId}/data/${page}`),
    dataSaverPageUrls: chapter.dataSaver.map((page) => `/api/pages/${chapterId}/data-saver/${page}`)
  };
}

File: data_engineer/producer/mangadex_client.py
from __future__ import annotations

import time
from typing import Any

import requests

from de_config import MangaDexConfig


USER_AGENT = "manga-data-engineering-demo/0.1"


class MangaDexClient:
    def __init__(self, config: MangaDexConfig) -> None:
        self.config = config
        self.session = requests.Session()
        self.session.headers.update({"Accept": "application/json", "User-Agent": USER_AGENT})

    def _get(self, path: str, params: list[tuple[str, str]] | None = None) -> dict[str, Any]:
        url = f"{self.config.base_url}{path}"
        for attempt in range(3):
            response = self.session.get(url, params=params, timeout=20)
            if response.status_code == 429 and attempt < 2:
                time.sleep(max(self.config.request_delay_seconds, 1.0) * (attempt + 1))
                continue
            response.raise_for_status()
            if self.config.request_delay_seconds > 0:
                time.sleep(self.config.request_delay_seconds)
            return response.json()
        raise RuntimeError(f"MangaDex request failed after retries: {url}")

    def list_manga(self, limit: int, offset: int) -> dict[str, Any]:
        params = [
            ("limit", str(limit)),
            ("offset", str(offset)),
            ("includes[]", "cover_art"),
            ("availableTranslatedLanguage[]", "en"),
            ("contentRating[]", "safe"),
            ("contentRating[]", "suggestive"),
            ("order[followedCount]", "desc"),
        ]
        return self._get("/manga", params)

    def list_chapters(self, manga_id: str, limit: int) -> dict[str, Any]:
        params = [
            ("limit", str(limit)),
            ("offset", "0"),
            ("translatedLanguage[]", "en"),
            ("includes[]", "scanlation_group"),
            ("order[volume]", "asc"),
            ("order[chapter]", "asc"),
        ]
        return self._get(f"/manga/{manga_id}/feed", params)


def first_localized(value: dict[str, str] | None, preferred: list[str] | None = None) -> str:
    if not value:
        return ""
    for language in preferred or ["en", "vi", "ja-ro", "ja"]:
        if value.get(language):
            return value[language]
    return next(iter(value.values()), "")


def slugify(value: str) -> str:
    return "".join(character.lower() if character.isalnum() else "-" for character in value).strip("-")


def normalize_manga(entity: dict[str, Any], popularity_rank: int) -> dict[str, Any]:
    attributes = entity.get("attributes") or {}
    relationships = entity.get("relationships") or []
    cover = next((item for item in relationships if item.get("type") == "cover_art"), None)
    file_name = ((cover or {}).get("attributes") or {}).get("fileName")
    tags = []

    for tag in attributes.get("tags") or []:
        tag_attributes = tag.get("attributes") or {}
        name = first_localized(tag_attributes.get("name"), ["en", "vi"])
        if not name:
            continue
        tags.append(
            {
                "mangadex_tag_id": tag.get("id"),
                "name": name,
                "slug": slugify(name),
                "group_name": tag_attributes.get("group"),
            }
        )

    alt_titles = [
        first_localized(title)
        for title in attributes.get("altTitles") or []
        if first_localized(title)
    ][:8]

    return {
        "mangadex_id": entity["id"],
        "title": first_localized(attributes.get("title")),
        "alt_titles": alt_titles,
        "description": first_localized(attributes.get("description")),
        "status": attributes.get("status"),
        "year": attributes.get("year"),
        "content_rating": attributes.get("contentRating"),
        "original_language": attributes.get("originalLanguage"),
        "publication_demographic": attributes.get("publicationDemographic"),
        "cover_url": f"https://uploads.mangadex.org/covers/{entity['id']}/{file_name}.512.jpg" if file_name else None,
        "popularity_rank": popularity_rank,
        "mangadex_created_at": attributes.get("createdAt"),
        "mangadex_updated_at": attributes.get("updatedAt"),
        "genres": tags,
    }


def normalize_chapter(entity: dict[str, Any], manga_db_id: int) -> dict[str, Any]:
    attributes = entity.get("attributes") or {}
    relationships = entity.get("relationships") or []
    group = next((item for item in relationships if item.get("type") == "scanlation_group"), None)
    return {
        "mangadex_id": entity["id"],
        "manga_id": manga_db_id,
        "title": attributes.get("title") or "",
        "chapter_number": attributes.get("chapter"),
        "volume": attributes.get("volume"),
        "translated_language": attributes.get("translatedLanguage") or "en",
        "pages": attributes.get("pages") or 0,
        "publish_at": attributes.get("publishAt"),
        "readable_at": attributes.get("readableAt"),
        "scanlation_group": ((group or {}).get("attributes") or {}).get("name"),
    }

